File: find_min_in_rotated_sorted_arr.py
# Time Complexity : O(log (n))
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : YES
# Any problem you faced while coding this : NO


from typing import List


class Solution:
    def findMin(self, nums: List[int]) -> int:
        if nums == None or not nums: return -1
        low, high = 0, len(nums) - 1
        while(low <= high):
            if(nums[low] <= nums[high]): return nums[low]
            mid = (low + high) // 2
            # to check if mid is smallest
            if ((nums[mid] == 0 or nums[mid] < nums[mid -1]) and (nums[mid] == len(nums) -1 or nums[mid] < nums[mid + 1])): 
                return nums[mid]
            elif nums[low] <= nums[mid]:
                low = mid + 1
            else: 
                high = mid - 1
                
sol = Solution()
print(sol.findMin([3]))

File: find_min_in_rotated_sorted_arr.java
//TC: O(log n)
class Solution {
    public int findMin(int[] nums) {
        
        //null check
        if(nums == null || nums.length == 0) return 0;
        
        int low = 0;
        int high = nums.length - 1;
        
        while(low <= high){
            if(nums[low] < nums[high]) return nums[low];
            int mid = low + (high-low)/2;
            
            if((mid == 0 || nums[mid] < nums[mid - 1]) && (mid == nums.length-1 || nums[mid] < nums[mid+1])){ //to check if mid element is the minimum element
                return nums[mid];
            }else if(nums[low] <= nums[mid] ){
                 low = mid+1;
            }else{
                high = mid-1;
            }
            
        }
        return -1;
        
    }
}
